datafusion/sql/src/planner.rs

```diff
@@ -2360,50 +2360,15 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             .is_ok()
     }
 
-    fn sql_array_literal(
-        &self,
-        elements: Vec<SQLExpr>,
-        schema: &DFSchema,
-    ) -> Result<Expr> {
-        let mut values = Vec::with_capacity(elements.len());
-
-        for element in elements {
-            let value =
-                self.sql_expr_to_logical_expr(element, schema, &mut HashMap::new())?;
-            match value {
-                Expr::Literal(scalar) => {
-                    values.push(scalar);
-                }
-                _ => {
-                    return Err(DataFusionError::NotImplemented(format!(
-                        "Arrays with elements other than literal are not supported: {}",
-                        value
-                    )));
-                }
-            }
-        }
-
-        let data_types: HashSet<DataType> =
-            values.iter().map(|e| e.get_datatype()).collect();
-
-        if data_types.is_empty() {
-            Ok(Expr::Literal(ScalarValue::List(
-                None,
-                Box::new(Field::new("item", DataType::Utf8, true)),
-            )))
-        } else if data_types.len() > 1 {
-            Err(DataFusionError::NotImplemented(format!(
-                "Arrays with different types are not supported: {:?}",
-                data_types,
-            )))
-        } else {
-            let data_type = values[0].get_datatype();
+    fn sql_array_expr(&self, elements: Vec<SQLExpr>, schema: &DFSchema) -> Result<Expr> {
+        let args: Vec<Expr> = elements
+            .into_iter()
+            .map(|expr| self.sql_expr_to_logical_expr(expr, schema, &mut HashMap::new()))
+            .collect::<Result<_>>()?;
 
-            Ok(Expr::Literal(ScalarValue::List(
-                Some(values),
-                Box::new(Field::new("item", data_type, true)),
-            )))
-        }
+        let fun = BuiltinScalarFunction::MakeArray;
+        // follow postgres convention and name result "array"
+        Ok(Expr::ScalarFunction { fun, args }.alias("array"))
```

Comment:
This is the rewrite in the planner to write `array[]` syntax to a `make_array` function call